File: packages/typescript/src/utils/rate-limiter.ts
export interface RateLimiterConfig {
  tokensPerSecond: number;
  burstSize: number;
  endpointOverrides?: Record<string, RateLimiterConfig>;
}

export interface RateLimiter {
  acquire(endpoint?: string): Promise<void>;
  tryAcquire(endpoint?: string): boolean;
  available(): number;
}

export class TokenBucketRateLimiter implements RateLimiter {
  private tokens: number;
  private lastRefill: number;
  private readonly maxTokens: number;
  private readonly refillRate: number;
  private readonly refillInterval: number;

  constructor(config: RateLimiterConfig) {
    this.maxTokens = config.burstSize;
    this.tokens = config.burstSize;
    this.lastRefill = Date.now();
    this.refillRate = config.tokensPerSecond;
    this.refillInterval = 1000;
  }

  private refill(): void {
    const now = Date.now();
    const elapsed = now - this.lastRefill;
    const tokensToAdd = (elapsed / this.refillInterval) * this.refillRate;
    this.tokens = Math.min(this.maxTokens, this.tokens + tokensToAdd);
    this.lastRefill = now;
  }

  tryAcquire(_endpoint?: string): boolean {
    this.refill();
    if (this.tokens >= 1) {
      this.tokens -= 1;
      return true;
    }
    return false;
  }

  async acquire(_endpoint?: string): Promise<void> {
    while (!this.tryAcquire()) {
      await new Promise((resolve) => setTimeout(resolve, 50));
    }
  }

  available(): number {
    this.refill();
    return this.tokens;
  }
}

export class NoopRateLimiter implements RateLimiter {
  acquire(_endpoint?: string): Promise<void> {
    return Promise.resolve();
  }

  tryAcquire(_endpoint?: string): boolean {
    return true;
  }

  available(): number {
    return Infinity;
  }
}

export class EndpointRateLimiterRouter implements RateLimiter {
  private readonly defaultLimiter: TokenBucketRateLimiter;
  private readonly endpointLimiters: Map<string, TokenBucketRateLimiter> = new Map();

  constructor(config: RateLimiterConfig) {
    this.defaultLimiter = new TokenBucketRateLimiter(config);
    if (config.endpointOverrides) {
      for (const [key, epConfig] of Object.entries(config.endpointOverrides)) {
        this.endpointLimiters.set(normalizeEndpointKey(key), new TokenBucketRateLimiter(epConfig));
      }
    }
  }

  private resolve(endpoint?: string): TokenBucketRateLimiter {
    if (!endpoint) return this.defaultLimiter;
    const key = normalizeEndpointKey(endpoint);
    const match = this.endpointLimiters.get(key);
    if (match) return match;
    for (const [pattern, limiter] of this.endpointLimiters) {
      if (key.startsWith(pattern)) return limiter;
    }
    return this.defaultLimiter;
  }

  acquire(endpoint?: string): Promise<void> {
    return this.resolve(endpoint).acquire();
  }

  tryAcquire(endpoint?: string): boolean {
    return this.resolve(endpoint).tryAcquire();
  }

  available(): number {
    return this.defaultLimiter.available();
  }

  availableFor(endpoint: string): number {
    return this.resolve(endpoint).available();
  }
}

export function normalizeEndpointKey(key: string): string {
  return key.toLowerCase().replace(/^https?:\/\//, "").replace(/\/+$/, "");
}

File: sdks/python/src/daraja/utils/rate_limiter.py
import threading
import time
from typing import Any, Protocol


class RateLimiter(Protocol):
    def acquire(self, endpoint: str | None = None) -> None: ...
    def try_acquire(self, endpoint: str | None = None) -> bool: ...


class TokenBucketRateLimiter:
    def __init__(self, tokens_per_second: float = 5.0, burst_size: int = 10) -> None:
        self._tokens_per_second = tokens_per_second
        self._burst_size = burst_size
        self._tokens = float(burst_size)
        self._last_refill = time.time()
        self._lock = threading.Lock()

    def acquire(self, endpoint: str | None = None) -> None:
        while not self.try_acquire(endpoint):
            time.sleep(0.01)

    def try_acquire(self, endpoint: str | None = None) -> bool:
        with self._lock:
            now = time.time()
            elapsed = now - self._last_refill
            self._tokens = min(self._burst_size, self._tokens + elapsed * self._tokens_per_second)
            self._last_refill = now

            if self._tokens >= 1.0:
                self._tokens -= 1.0
                return True
            return False


class NoopRateLimiter:
    def acquire(self, endpoint: str | None = None) -> None:
        pass

    def try_acquire(self, endpoint: str | None = None) -> bool:
        return True


class RateLimiterConfig:
    def __init__(self, tokens_per_second: float = 5.0, burst_size: int = 10,
                 endpoint_overrides: dict[str, dict[str, Any]] | None = None) -> None:
        self.tokens_per_second = tokens_per_second
        self.burst_size = burst_size
        self.endpoint_overrides = endpoint_overrides or {}


class EndpointRateLimiterRouter:
    def __init__(self, config: RateLimiterConfig) -> None:
        self._default = TokenBucketRateLimiter(
            tokens_per_second=config.tokens_per_second,
            burst_size=config.burst_size,
        )
        self._endpoint_limiters: dict[str, TokenBucketRateLimiter] = {}
        for key, ep_cfg in (config.endpoint_overrides or {}).items():
            self._endpoint_limiters[_normalize_key(key)] = TokenBucketRateLimiter(
                tokens_per_second=ep_cfg.get("tokens_per_second", config.tokens_per_second),
                burst_size=ep_cfg.get("burst_size", config.burst_size),
            )

    def acquire(self, endpoint: str | None = None) -> None:
        self._resolve(endpoint).acquire()

    def try_acquire(self, endpoint: str | None = None) -> bool:
        return self._resolve(endpoint).try_acquire()

    def _resolve(self, endpoint: str | None) -> TokenBucketRateLimiter:
        if not endpoint:
            return self._default
        key = _normalize_key(endpoint)
        if key in self._endpoint_limiters:
            return self._endpoint_limiters[key]
        for pattern, limiter in self._endpoint_limiters.items():
            if key.startswith(pattern):
                return limiter
        return self._default


def _normalize_key(key: str) -> str:
    return key.lower().replace("https://", "").replace("http://", "").rstrip("/")
